File: problem-types/field.py
def entry_form(data, kwargs):
	mask = data['mask'].split('\n')
	assert len(mask) == data['height']

	yield '<table class="grid field">'
	for y, row in enumerate(mask):
		assert len(row) == data['width']
		yield '<tr>'
		for x, cell in enumerate(row):
			yield '<td class="tile">' if cell == 'x' else '<td>'
	yield '</table>'

def validate(data, answer):
	answer = '\n'.join(answer.split())
	return answer in data['corrects']

File: problem-types/field.js
"use strict";

const checked = 'checked';
const base = document.querySelector('.grid');
const sender = document.createElement('input');
sender.type = 'hidden';
sender.name = 'answer';
base.parentNode.appendChild(sender);

base.addEventListener('click', function(e) {
	let tile = e.target.closest('.tile');
	if (!tile || !base.contains(tile))
		return;
	tile.classList.toggle('checked');

	let answer = [];
	for (const row of base.rows) {
		if (row.classList.contains('pad'))
			continue;
		let answer_row = [];
		for (const cell of row.cells) {
			if (cell.classList.contains('pad'))
				continue;
			answer_row.push(cell.classList.contains(checked) ? 'x' : '-');
		}
		answer.push(answer_row.join(''))
	}
	sender.value = answer.join('\n')
});

